src/libra_py/dynamics/qtag/QTAG_main.py

```diff
@@ -0,0 +1,229 @@
+	if coupling_name == 'LHA' or coupling_name == 'BAT':
+		cplg=getattr(QTAG_ham,coupling_name)
+	else:
+		cplg=getattr(QTAG_pots,coupling_name)
+except AttributeError:
+	print("Potential Model: ",model_name)
+	print("Coupling: ",coupling_name)
+	sys.exit("Error in model dictionary, check that pot_type and rep are compatible!")
+
+try:
+	mom_calc=getattr(QTAG_mom,str(mom_params['adjust']))
+except AttributeError:
+	sys.exit("Error in mom_params dictionary: 'adjust' keyword not recognized!")
+
+try:
+	propagate=getattr(QTAG_prop,str(mss['prop_method']))
+except AttributeError:
+	sys.exit("Error in mss dictionary: 'prop_method' keyword not recognized!")
+
+#Rename variables locally for convenience...
+ndof,ntraj,niter,dt=univ['ndof'],univ['ntraj'],univ['niter'],univ['dt']
+n_data_out,n_snapshots=univ['n_data_out'],univ['n_snapshots']
+n_wf_print=int(niter/n_snapshots)
+
+#Define MATRIX and CMATRIX objects...
+b1=CMATRIX(ntraj,1);b2=CMATRIX(ntraj,1)
+c1_new=CMATRIX(ntraj,1);c2_new=CMATRIX(ntraj,1)
+ct_old=CMATRIX(2*ntraj,1);ct_new=CMATRIX(2*ntraj,1)
+dummy=CMATRIX(ntraj,ntraj)
+
+#Create list used in nonad vector assignments...
+pop_list=[]
+for i in range(ntraj):
+	pop_list.append(i)
+
+#Initialize the basis parameters {q,p,a,s} and create initial b-matrix...
+try:
+	qpas1=getattr(QTAG_init,str(traj0['placement']))(ndof,ntraj,traj0,wf0)
+	qpas2=getattr(QTAG_init,str(traj0['placement']))(ndof,ntraj,traj0,wf0)
+except AttributeError:
+	sys.exit("Error in traj0 dictionary: 'placement' keyword not recognized!")
+
+#Create initial projection vectors...
+b1=QTAG_init.coeffs(ndof,ntraj,wf0,qpas1,1)
+b2=QTAG_init.coeffs(ndof,ntraj,wf0,qpas2,2)
+bt=QTAG_ham.nonad_assemble("cplx",ntraj,1,b1,b2,dummy)
+
+#Begin the iterative process...
+t=0.0
+t_start=time.perf_counter()
+for iter in range(niter):
+	if iter%(niter/20)==0:
+		print("Iteration = ",iter, " of ", niter)
+
+	ov1=QTAG_calc.overlap(qpas1,qpas1)
+	ov2=QTAG_calc.overlap(qpas2,qpas2)
+	ov12=QTAG_calc.overlap(qpas1,qpas2)
+	
+	H1=QTAG_ham.hamiltonian(qpas1,ov1,1,vcalc,pot)
+	H2=QTAG_ham.hamiltonian(qpas2,ov2,2,vcalc,pot)
+	Hcpl=QTAG_ham.coupling(qpas1,qpas2,ov12,3,cplg,pot)
+
+	ovt=QTAG_ham.nonad_assemble("cplx",ntraj,ntraj,ov1,ov2,dummy)
+	Ht=QTAG_ham.nonad_assemble("cplx",ntraj,ntraj,H1,H2,Hcpl)
+
+	ct_new=QTAG_ham.basis_diag(2*ntraj,Ht,ovt,bt)
+	pop_submatrix(ct_new,c1_new,pop_list,[0])
+	pop_submatrix(ct_new,c2_new,[ntraj+i for i in pop_list],[0])
+
+	norm1=QTAG_calc.norm(c1_new,ov1)
+	norm2=QTAG_calc.norm(c2_new,ov2)
+	energy=QTAG_calc.energy(ct_new,Ht)
+	
+	qpas1n,qpas2n,b1,b2=propagate(mom_calc,props,qpas1,c1_new,qpas2,c2_new,norm2)
+
+	for vals in range(4):
+		qpas1[vals]=QTAG_calc.update(ndof,qpas1[vals],qpas1n[vals])
+		qpas2[vals]=QTAG_calc.update(ndof,qpas2[vals],qpas2n[vals])
+
+	bt=QTAG_ham.nonad_assemble("cplx",ntraj,1,b1,b2,dummy)
+#	ct_old=QTAG_calc.update(1,ct_old,ct_new)
+
+	norm12=np.abs((c1_new.T().conj()*ov12*c2_new).get(0))**2
+
+	if iter%n_data_out==0:
```

Comment:
just overall - can you make the code structure to be the same for any number of DOFs? the specific code would do things based on the actual number of DOFs - e.g. you could do a loop over arbitrary "ndof" iteractions and do the same operation (e.g. writing) in each iteration (of course, the specific parameters would be different)